extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/SegmentLoadWaiter.java

```diff
@@ -0,0 +1,317 @@
+import org.jboss.netty.handler.codec.http.HttpMethod;
+import org.joda.time.DateTime;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import javax.ws.rs.core.MediaType;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class that periodically checks with the broker if all the segments generated are loaded by querying the sys table
+ * and blocks till it is complete. This will account for and not wait for segments that would never be loaded due to
+ * load rules. Should only be called if the query generates new segments or tombstones.
+ * <br>
+ * If an exception is thrown during operation, this will simply log the exception and exit without failing the task,
+ * since the segments have already been published successfully, and should be loaded eventually.
+ * <br>
+ * If the segments are not loaded within {@link #TIMEOUT_DURATION_MILLIS} milliseconds, this logs a warning and exits
+ * for the same reason.
+ */
+public class SegmentLoadWaiter
+{
+  private static final Logger log = new Logger(SegmentLoadWaiter.class);
+  private static final long INITIAL_SLEEP_DURATION_MILLIS = TimeUnit.SECONDS.toMillis(5);
+  private static final long SLEEP_DURATION_MILLIS = TimeUnit.SECONDS.toMillis(5);
+  private static final long TIMEOUT_DURATION_MILLIS = TimeUnit.MINUTES.toMillis(10);
+  private static final String LOAD_QUERY = "SELECT COUNT(*) AS totalSegments,\n"
+                                           + "COUNT(*) FILTER (WHERE is_available = 0 AND is_published = 1 AND replication_factor != 0) AS loadingSegments\n"
+                                           + "FROM sys.segments\n"
+                                           + "WHERE datasource = '%s' AND is_overshadowed = 0 AND version = '%s'";
+
+  private final BrokerClient brokerClient;
+  private final ObjectMapper objectMapper;
+  // Map of version vs latest load status.
+  private final Map<String, VersionLoadStatus> versionToLoadStatusMap;
+  private final String datasource;
+  private final Set<String> versionsToAwait;
+  private final boolean doWait;
+  private volatile SegmentLoadWaiterStatus status;
+
+  public SegmentLoadWaiter(ControllerContext context, String datasource, Set<String> versionsToAwait, int initialSegmentCount)
+  {
+    this.brokerClient = context.injector().getInstance(BrokerClient.class);
+    this.objectMapper = context.jsonMapper();
+    this.datasource = datasource;
+    this.versionsToAwait = new TreeSet<>(versionsToAwait);
+    this.versionToLoadStatusMap = new HashMap<>();
+    this.status = new SegmentLoadWaiterStatus(State.INIT, null, 0, initialSegmentCount, initialSegmentCount);
+    this.doWait = true;
+  }
+
+  @VisibleForTesting
+  SegmentLoadWaiter(BrokerClient brokerClient, ObjectMapper objectMapper, String datasource, Set<String> versionsToAwait, int initialSegmentCount, boolean doWait)
```

Comment:
```suggestion
  public SegmentLoadWaiter(BrokerClient brokerClient, ObjectMapper objectMapper, String datasource, Set<String> versionsToAwait, int initialSegmentCount, boolean doWait)
```
The public constructor should be removed, the constructor that's visible for testing should be made public, and the `ControllerImpl` should directly call the constructor with the extracted values instead, otherwise looks like we are duplicating a lot of work here.